source/fundamentals/crud/write-operations/pkFactory.txt

```diff
@@ -0,0 +1,99 @@
+.. _node-pkfactory:
+
+===================
+Primary Key Factory
+===================
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn about the primary key factory features
+and limitations in the MongoDB Node.js driver. 
+
+Behavior
+--------
+
+A primary key factory is a class that defines a system for creating
+unique identifiers for documents and implements  the ``pkFactory``
+interface. 
+
+The driver uses the primary key factory to generate unique ``_id``
+values for your documents when you choose not to specify an ``_id``
+during an :ref:`insert operation <node-fundamentals-insert-data>`. The
+default primary key factory generates ``ObjectId`` values.
+
+Upsert Operations
+~~~~~~~~~~~~~~~~~
+
+For :ref:`upsert operations <node-fundamentals-upsert>`, the driver
+doesn't use any ``pkFactory`` methods. Instead, the driver always
+creates ``ObjectId`` values for the ``_id`` field of the
+upserted documents.
+
+If you want to use ``pkFactory`` methods, peform a :ref:`find
+operation <node-fundamentals-retrieve-data>`, then an :ref:`update
+<node-fundamentals-change-a-document>` or :ref:`insert
+<node-fundamentals-insert-data>` operation.
+
+Data Consistency
+----------------
+
+If you do not specify a primary key factory method that generates
+``_id`` values of your desired type, you may experience data
+inconsistency.
+
+This may occur because the driver's default primary factory method generates
+``_id`` values of type ``ObjectId`` regardless of the type you specify for the 
+``_id`` field in your type parameter.
+
+Example
+~~~~~~~
+
+Assume you apply the following interface to your ``Collection``
+instance:
+
+.. code-block:: typescript
+
+   interface I {
+     _id?: number;
+     name: string;
+   } 
+
+The following insert operations insert a document with an ``_id`` of
+type ``number`` and another document with an ``_id`` of type
+``ObjectId`` into your collection:
+
+.. code-block:: typescript
+
+   collection.insertOne({ _id: 1, name: "Scar" });
+   collection.insertOne({ name: "Ursula" });
+
+To keep the ``_id`` consistent as whole numbers, apply a primary key
+factory to your ``MongoClient`` instance like this:
+
+.. code-block:: typescript
+
+   const client = new MongoClient(uri, {
+     pkFactory: { createPk: () => Math.floor(Math.random() * Number.MAX_SAFE_INTEGER) },
+   });
```

Comment:
issue: There are a few minor issues here.

1. I expected to see this in a top level heading for "Specify your own `pkFactory`. This is a (mostly) great example and it shouldn't be hidden since it illustrates how to specify a pkFactory, and why you would.
2. Rather than "whole number", let's use `Integer`.
3. But! Instead of using an integer here (which is a very bad _id scheme), how about [UUID](https://www.npmjs.com/package/uuid) instead, which is a nice alternative.